File: content/examples/other/leetcode/problems/longest-substring-without-repeating-characters/example-2/solution.js
function lengthOfLongestSubstring(s) {
  const last = new Map();
  let left = 0;
  let best = 0;
  for (let right = 0; right < s.length; right += 1) {
    const ch = s[right];
    if (last.has(ch)) left = Math.max(left, last.get(ch) + 1);
    last.set(ch, right);
    best = Math.max(best, right - left + 1);
  }
  return best;
}

if (require.main === module) {
  console.log(lengthOfLongestSubstring("abcabcbb"));
}

module.exports = { lengthOfLongestSubstring };

File: content/examples/other/leetcode/problems/longest-substring-without-repeating-characters/example-2/Solution.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
  static int lengthOfLongestSubstring(String s) {
    Map<Character, Integer> last = new HashMap<>();
    int left = 0;
    int best = 0;
    for (int right = 0; right < s.length(); right += 1) {
      char ch = s.charAt(right);
      if (last.containsKey(ch)) left = Math.max(left, last.get(ch) + 1);
      last.put(ch, right);
      best = Math.max(best, right - left + 1);
    }
    return best;
  }

  public static void main(String[] args) {
    System.out.println(lengthOfLongestSubstring("abcabcbb"));
  }
}
